File: dsa/week(12)/binarytree.js
class BinaryTreeNode {
  constructor(data) {
    this.data = data;
    this.left = null;
    this.right = null;
  }
}

function printTree(root) {
  if (root === null) {
    return;
  }
  process.stdout.write(root.data + ":");
  if (root.left !== null) {
    process.stdout.write("L" + root.left.data);
  }
  if (root.right !== null) {
    process.stdout.write("R" + root.right.data);
  }
  console.log();
  printTree(root.left);
  printTree(root.right);
}

function takeInput() {
  let rootData = readline();
  if (rootData === "-1") {
    return null;
  }
  let root = new BinaryTreeNode(parseInt(rootData));
  root.left = takeInput();
  root.right = takeInput();
  return root;
}

// Function to calculate the number of nodes in the binary tree
function numNodes(root) {
  if (root === null) {
    return 0;
  }
  return 1 + numNodes(root.left) + numNodes(root.right);
}

// Function to take input for the binary tree level-wise
function takeInputLevelWise() {
  let rootData = readline();
  if (rootData === "-1") {
    return null;
  }
  let root = new BinaryTreeNode(parseInt(rootData));
  let queue = [];
  queue.push(root);
  while (queue.length !== 0) {
    let front = queue.shift();
    let leftChildData = readline();
    if (leftChildData !== "-1") {
      let child = new BinaryTreeNode(parseInt(leftChildData));
      front.left = child;
      queue.push(child);
    }
    let rightChildData = readline();
    if (rightChildData !== "-1") {
      let child = new BinaryTreeNode(parseInt(rightChildData));
      front.right = child;
      queue.push(child);
    }
  }
  return root;
}

// Function to perform inorder traversal of the binary tree
function inorder(root) {
  if (root === null) {
    return;
  }
  inorder(root.left);
  process.stdout.write(root.data + " ");
  inorder(root.right);
}

// Function to perform preorder traversal of the binary tree
function preorder(root) {
  if (root === null) {
    return;
  }
  process.stdout.write(root.data + " ");
  preorder(root.left);
  preorder(root.right);
}

// Function to perform postorder traversal of the binary tree
function postorder(root) {
  if (root === null) {
    return;
  }
  postorder(root.left);
  postorder(root.right);
  process.stdout.write(root.data + " ");
}

// Function to build the binary tree from inorder and preorder arrays
function buildTreeHelper(inorder, preorder, inStart, inEnd, preStart, preEnd) {
  if (inStart > inEnd) {
    return null;
  }
  let rootIndex = -1;
  let rootData = preorder[preStart];
  for (let i = inStart; i <= inEnd; i++) {
    if (inorder[i] === rootData) {
      rootIndex = i;
      break;
    }
  }
  let leftInStart = inStart;
  let leftInEnd = rootIndex - 1;
  let leftPreStart = preStart + 1;
  let leftPreEnd = leftInEnd - leftInStart + leftPreStart;

  let rightPreStart = leftPreEnd + 1;
  let rightPreEnd = preEnd;
  let rightInStart = rootIndex + 1;
  let rightInEnd = inEnd;

  let root = new BinaryTreeNode(rootData);
  root.left = buildTreeHelper(
    inorder,
    preorder,
    leftInStart,
    leftInEnd,
    leftPreStart,
    leftPreEnd
  );
  root.right = buildTreeHelper(
    inorder,
    preorder,
    rightInStart,
    rightInEnd,
    rightPreStart,
    rightPreEnd
  );
  return root;
}

// Function to build the binary tree from inorder and preorder arrays
function buildTree(inorder, preorder, size) {
  return buildTreeHelper(inorder, preorder, 0, size - 1, 0, size - 1);
}

// Function to calculate the height of the binary tree
function height(root) {
  if (root === null) {
    return 0;
  }
  return 1 + Math.max(height(root.left), height(root.right));
}

// Function to calculate the diameter of the binary tree
function diameterOfABinaryTree(root) {
  if (root === null) {
    return 0;
  }
  let option1 = height(root.left) + height(root.right) + 1;
  let option2 = diameterOfABinaryTree(root.left);
  let option3 = diameterOfABinaryTree(root.right);
  return Math.max(option1, Math.max(option2, option3));
}

// Function to calculate the height and diameter of the binary tree
function heightDiameter(root) {
  if (root === null) {
    return { height: 0, diameter: 0 };
  }
  let leftAns = heightDiameter(root.left);
  let rightAns = heightDiameter(root.right);
  let height = 1 + Math.max(leftAns.height, rightAns.height);
  let diameter = Math.max(
    height,
    Math.max(leftAns.diameter, rightAns.diameter),
    leftAns.height + rightAns.height + 1
  );
  return { height, diameter };
}

// Example usage
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];

rl.on("line", (line) => {
  input.push(line);
}).on("close", () => {
  let root = takeInputLevelWise();
  printTree(root);
  console.log("Inorder traversal:");
  inorder(root);
  console.log("\nPreorder traversal:");
  preorder(root);
  console.log("\nPostorder traversal:");
  postorder(root);
  console.log("\nHeight of the tree:", height(root));
  console.log("Diameter of the tree:", diameterOfABinaryTree(root));
  let { height: treeHeight, diameter: treeDiameter } = heightDiameter(root);
  console.log("Height and diameter of the tree:", treeHeight, treeDiameter);
});

File: dsa/week(12)/binarytree.py

from collections import deque
from typing import Optional, Tuple

class BinaryTreeNode:
    def __init__(self, data: int):
        self.data = data
        self.left = None
        self.right = None

    def __del__(self):
        if self.left:
            del self.left
        if self.right:
            del self.right

def print_tree(root: Optional[BinaryTreeNode]) -> None:
    if root is None:
        return
    print(f"{root.data}:", end="")
    if root.left:
        print(f"L{root.left.data}", end="")
    if root.right:
        print(f"R{root.right.data}", end="")
    print()
    print_tree(root.left)
    print_tree(root.right)

def take_input() -> Optional[BinaryTreeNode]:
    root_data = int(input("Enter data\n"))
    if root_data == -1:
        return None
    root = BinaryTreeNode(root_data)
    left_child = take_input()
    right_child = take_input()
    root.left = left_child
    root.right = right_child
    return root

def num_nodes(root: Optional[BinaryTreeNode]) -> int:
    if root is None:
        return 0
    return 1 + num_nodes(root.left) + num_nodes(root.right)

def take_input_level_wise() -> Optional[BinaryTreeNode]:
    root_data = int(input("Enter root data\n"))
    if root_data == -1:
        return None
    root = BinaryTreeNode(root_data)
    pending_nodes = deque([root])
    while pending_nodes:
        front = pending_nodes.popleft()
        left_child_data = int(input(f"Enter left child of {front.data}\n"))
        if left_child_data != -1:
            child = BinaryTreeNode(left_child_data)
            front.left = child
            pending_nodes.append(child)
        right_child_data = int(input(f"Enter right child of {front.data}\n"))
        if right_child_data != -1:
            child = BinaryTreeNode(right_child_data)
            front.right = child
            pending_nodes.append(child)
    return root

def inorder(root: Optional[BinaryTreeNode]) -> None:
    if root is None:
        return
    inorder(root.left)
    print(root.data, end=" ")
    inorder(root.right)

def preorder(root: Optional[BinaryTreeNode]) -> None:
    if root is None:
        return
    print(root.data, end=" ")
    preorder(root.left)
    preorder(root.right)

def postorder(root: Optional[BinaryTreeNode]) -> None:
    if root is None:
        return
    postorder(root.left)
    postorder(root.right)
    print(root.data, end=" ")

def build_tree_helper(inorder: list[int], preorder: list[int], in_start: int, in_end: int, pre_start: int, pre_end: int) -> Optional[BinaryTreeNode]:
    if in_start > in_end:
        return None
    root_data = preorder[pre_start]
    root_index = inorder.index(root_data)
    left_in_start = in_start
    left_in_end = root_index - 1
    left_pre_start = pre_start + 1
    left_pre_end = left_pre_start + (left_in_end - left_in_start)
    right_pre_start = left_pre_end + 1
    right_pre_end = pre_end
    right_in_start = root_index + 1
    right_in_end = in_end
    root = BinaryTreeNode(root_data)
    root.left = build_tree_helper(inorder, preorder, left_in_start, left_in_end, left_pre_start, left_pre_end)
    root.right = build_tree_helper(inorder, preorder, right_in_start, right_in_end, right_pre_start, right_pre_end)
    return root

def build_tree(inorder: list[int], preorder: list[int], size: int) -> Optional[BinaryTreeNode]:
    return build_tree_helper(inorder, preorder, 0, size - 1, 0, size - 1)

def height(root: Optional[BinaryTreeNode]) -> int:
    if root is None:
        return 0
    return 1 + max(height(root.left), height(root.right))

def diameter_of_a_binary_tree(root: Optional[BinaryTreeNode]) -> int:
    if root is None:
        return 0
    option1 = height(root.left) + height(root.right) + 1
    option2 = diameter_of_a_binary_tree(root.left)
    option3 = diameter_of_a_binary_tree(root.right)
    return max(option1, option2, option3)

def height_diameter(root: Optional[BinaryTreeNode]) -> Tuple[int, int]:
    if root is None:
        return (0, 0)
    left_ans = height_diameter(root.left)
    right_ans = height_diameter(root.right)
    current_height = 1 + max(left_ans[0], right_ans[0])
    current_diameter = max(left_ans[1], right_ans[1], left_ans[0] + right_ans[0] + 1)
    return (current_height, current_diameter)
